src/path/dec_int.rs

```diff
@@ -36,23 +37,61 @@ use {core::fmt, std::ffi::OsStr, std::path::Path};
 /// ```
 #[derive(Clone)]
 pub struct DecInt {
-    // 20 `u8`s is enough to hold the decimal ASCII representation of any
-    // `u64`, and we add one for a NUL terminator for `as_c_str`.
-    buf: [u8; 20 + 1],
+    // Enough to hold an i64 and NUL terminator.
+    buf: [MaybeUninit<u8>; "-9223372036854775808\0".len()],
     len: usize,
 }
 
 impl DecInt {
     /// Construct a new path component from an integer.
     #[inline]
-    pub fn new<Int: Integer>(i: Int) -> Self {
-        let mut me = DecIntWriter(Self {
-            buf: [0; 20 + 1],
-            len: 0,
-        });
-        let mut buf = Buffer::new();
-        me.write_str(buf.format(i)).unwrap();
-        me.0
+    pub fn new<Int: Integer + 'static>(i: Int) -> Self {
+        let mut buf = [MaybeUninit::uninit(); 21];
+
+        let mut str_buf = Buffer::new();
+        let str_buf = str_buf.format(i);
+        {
+            let max_buf_size = {
+                let bits = match TypeId::of::<Int>() {
+                    id if [TypeId::of::<i8>(), TypeId::of::<u8>()].contains(&id) => u8::BITS,
+                    id if [TypeId::of::<i16>(), TypeId::of::<u16>()].contains(&id) => u16::BITS,
+                    id if [TypeId::of::<i32>(), TypeId::of::<u32>()].contains(&id) => u32::BITS,
+                    id if [TypeId::of::<i64>(), TypeId::of::<u64>()].contains(&id) => u64::BITS,
+                    id if [TypeId::of::<i128>(), TypeId::of::<u128>()].contains(&id) => u128::BITS,
+                    id if [TypeId::of::<isize>(), TypeId::of::<usize>()].contains(&id) => {
+                        usize::BITS
+                    }
+                    _ => unreachable!(),
+                };
+                match bits {
+                    8 => "-128".len(),
+                    16 => "-32768".len(),
+                    32 => "-2147483648".len(),
+                    64 => "-9223372036854775808".len(),
+                    128 => "-170141183460469231731687303715884105728".len(),
+                    _ => unreachable!(),
+                }
+            };
+            if str_buf.len() > max_buf_size {
+                unsafe { core::hint::unreachable_unchecked() }
+            }
```

Comment:
Not quite, the assertion is a bit of a red herring. It's there to make the error message nicer on {u,i}128s because otherwise we panic on bounds check failures. If you remove the hint, you'll get bounds check panics on all ints (or rather the assertion will be compiled in, but if we made it debug_assert then release mode would see bounds checks).

It's probably easier to see this by playing with https://rust.godbolt.org/z/dqb53Pj8o. Change the param in the function foo at the top to see how the different monomorphizations look.

Reply:
Ah, thanks, I get it now. I played around with the link a little, and came up with this: https://rust.godbolt.org/z/4TEvYc4x8. It codegens to the same code, but does the assert before the unreachable hint so that we don't invoke UB before the assert in debug builds, and it doesn't use the `TypeId` checks. What do you think?

Reply:
Unfortunately that has UB if you pass in a `u128::MAX` for example (str_buf will be 40 long while buf is only 21). For debug builds they should already be getting an assertion as long as they don't enable optimizations: https://doc.rust-lang.org/src/core/hint.rs.html#100. One idea is to change the TypeId stuff to panic on 128 bit values so that we can use your simplified hint condition, but that'll still have very ugly TypeId checks.